Userland/Libraries/LibGfx/ImageFormats/JPEG2000ProgressionIterators.cpp

```diff
@@ -9,23 +9,30 @@
 namespace Gfx::JPEG2000 {
 
 LayerResolutionLevelComponentPositionProgressionIterator::LayerResolutionLevelComponentPositionProgressionIterator(int layer_count, int max_number_of_decomposition_levels, int component_count, Function<int(int resolution_level, int component)> precinct_count)
-    : m_precinct_count(move(precinct_count))
+    : m_layer_count(layer_count)
+    , m_max_number_of_decomposition_levels(max_number_of_decomposition_levels)
+    , m_component_count(component_count)
+    , m_precinct_count(move(precinct_count))
+    , m_generator(generator())
+
 {
-    m_end.layer = layer_count;
-    m_end.resolution_level = max_number_of_decomposition_levels + 1;
-    m_end.component = component_count;
-    m_end.precinct = m_precinct_count(m_next.resolution_level, m_next.component);
+    m_next = m_generator.next();
 }
 
 bool LayerResolutionLevelComponentPositionProgressionIterator::has_next() const
 {
-    return m_next != ProgressionData { m_end.layer, 0, 0, 0 };
+    return m_next.has_value();
 }
 
 ProgressionData LayerResolutionLevelComponentPositionProgressionIterator::next()
 {
-    ProgressionData current_data = m_next;
+    auto result = m_next;
+    m_next = m_generator.next();
+    return result.value();
```

Comment:
I have the feeling this is a common pattern when forwarding iterators/generators, maybe we should come up with an idiom
(Not now though)